README.md
# Blinkt

[![Build Status](https://travis-ci.com/golemparts/blinkt.svg?branch=master)](https://travis-ci.com/golemparts/blinkt)
[![crates.io](https://img.shields.io/crates/v/blinkt)](https://crates.io/crates/blinkt)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![Minimum rustc version](https://img.shields.io/badge/rustc-v1.56.0-lightgray.svg)](https://blog.rust-lang.org/2021/10/21/Rust-1.56.0.html)

Blinkt is a Rust library that provides an interface for the Pimoroni Blinkt!, and any similar APA102 or SK9822 LED strips or boards, on a Raspberry Pi. The library supports bitbanging mode on any GPIO pins, and hardware SPI mode on GPIO 10 (physical pin 19) for data, and GPIO 11 (physical pin 23) for clock.

For bitbanging mode, Blinkt gains access to the BCM283x GPIO peripheral either through `/dev/gpiomem` or `/dev/mem`. Hardware SPI mode is controlled through `/dev/spidev0.0`.

Both the original APA102 and the SK9822 clone are supported. The RGB LED/driver ICs are referred to as pixels throughout the code and documentation.

Backwards compatibility for minor revisions isn't guaranteed until the library reaches v1.0.0.

Blinkt is under development on the [master branch](https://github.com/golemparts/blinkt/tree/master) of the repository on GitHub. If you're looking for the `README.md` or the `examples` directory for the latest release or any of the earlier releases, visit [crates.io](https://crates.io/crates/blinkt), download an archived release from the GitHub [releases](https://github.com/golemparts/blinkt/releases) page, or clone and checkout the relevant release tag.

## Documentation

Online documentation is available for the latest release, older releases, and the version currently in development.

* Latest release: [docs.golemparts.com/blinkt](https://docs.golemparts.com/blinkt)
* Older releases: [docs.rs/blinkt](https://docs.rs/blinkt)
* In development: [docs.golemparts.com/blinkt-dev](https://docs.golemparts.com/blinkt-dev)

## Usage

Add a dependency for `blinkt` to your `Cargo.toml` using `cargo add blinkt`, or by adding the following line to your dependencies section.

```toml
[dependencies]
blinkt = "0.7.1"
```

Call `Blinkt::new()` to create a new Blinkt with the default settings. Alternative configuration options are available through `Blinkt::with_settings()` and `Blinkt::with_spi()`.

```rust
use blinkt::Blinkt;

let mut blinkt = Blinkt::new()?;
```

## Examples

The example below demonstrates swapping all pixels on a Blinkt! board between red, green and blue.

```rust
use std::error::Error;
use std::time::Duration;
use std::{mem, thread};

use blinkt::Blinkt;

fn main() -> Result<(), Box<dyn Error>> {
    let mut blinkt = Blinkt::new()?;
    let (red, green, blue) = (&mut 255, &mut 0, &mut 0);

    loop {
        blinkt.set_all_pixels(*red, *green, *blue);
        blinkt.show()?;

        thread::sleep(Duration::from_millis(250));

        mem::swap(red, green);
        mem::swap(red, blue);
    }
}
```

To control an LED strip consisting of 144 pixels, connected to the Raspberry Pi's hardware SPI pins (data on GPIO 10 (physical pin 19), and clock on GPIO 11 (physical pin 23)), at 16 MHz clock speed, replace the `Blinkt::new()` line in the above example with the following. You may have to tweak the maximum clock speed based on the number of pixels and the wire quality.

```rust
let mut blinkt = Blinkt::with_spi(16_000_000, 144)?;
```

Additional examples can be found in the `examples` directory.

## Cross compilation

If you're not working directly on a Raspberry Pi, you'll have to cross-compile your code for the appropriate ARM architecture. Check out [this guide](https://github.com/japaric/rust-cross) for more information, or try the [cross](https://github.com/japaric/cross) project for "zero setup" cross compilation.

### Cargo

While additional steps may be necessary to cross-compile binaries on your platform, checking your code with `cargo check` only requires the installation of an appropriate target. Most Raspberry Pi models need the `armv7-unknown-linux-gnueabihf` target. For some models, like the Raspberry Pi Zero, a different target triple is required.

Install the relevant target using `rustup`.

```bash
rustup target install armv7-unknown-linux-gnueabihf
```

In the root directory of your project, create a `.cargo` subdirectory, and then save the following snippet to `.cargo/config`.

```toml
[build]
target = "armv7-unknown-linux-gnueabihf"
```

### Visual Studio Code

The rust-analyzer extension for Visual Studio Code needs to be made aware of the target platform by setting the `rust-analyzer.cargo.target` configuration option. In the root directory of your project, create a `.vscode` subdirectory, and then save the following snippet to `.vscode/settings.json`.

```json
{
    "rust-analyzer.cargo.target": "armv7-unknown-linux-gnueabihf"
}
```

## Copyright and license

Copyright (c) 2016-2023 Rene van der Meer. Released under the [MIT license](LICENSE).


CHANGELOG.md
# Changelog

## 0.7.1 (Oct 18, 2023)

* Upgrade `rppal` dependency to 0.15.0.

## 0.7.0 (Aug 14, 2022)

* (Breaking change) Transition to Rust 2021, requiring rustc v1.56.0 or newer to compile the library.
* Fix display issues under high CPU load (contributed by @ashhhleyyy).
* Change iterator type to `slice::IterMut` (contributed by @hellow554).
* Extend SPI interface with bus, SS and mode selection (contributed by @hellow554).
* Upgrade `rppal` dependency to 0.13.1.

## 0.6.0 (May 12, 2021)

* (Breaking change) Transition to Rust 2018, requiring rustc v1.45.0 or newer to compile the library.
* Add new badge to `README.md`, indicating the required minimum rustc version.
* Upgrade `rppal` dependency to 0.12.0.
* Add support for Raspberry Pi CM3+, CM4, 4B, 400.
* Remove `quick-error` dependency.
* Add `Pixel` to public interface, which represents a pixel on an LED strip or board.
* Add `IterMut`, which implements `Iterator` for a `&mut Pixel` slice.
* Implement `IntoIterator` for `&mut Blinkt`, which returns an `IterMut` mutable iterator over all `Pixel`s stored in `Blinkt`.
* Add `Blinkt::iter_mut()`, which returns an `IterMut` mutable iterator over all `Pixel`s stored in `Blinkt`.
* Add `Blinkt::clear_on_drop()`, which returns the current value of `clear_on_drop`.
* (Breaking change) Remove `Blinkt::cleanup()`. When `Blinkt` goes out of scope, any changed pin states are automatically reset. If `clear_on_drop` is set to `true`, all pixels will also be cleared.
* Implement `Send` for `Blinkt`.

## 0.5.0 (November 16, 2018)

* Add support for Raspberry Pi 3 A+.
* Replace `spidev` dependency with `rppal`'s SPI module.
* (Breaking change) Add `Error::Spi` to indicate an SPI error occurred.

## 0.4.0 (April 21, 2018)

* Add support for hardware SPI through `Blinkt::with_spi()`.
* (Breaking change) Add a `Result` return value to `Blinkt::show()` to catch potential SPI or GPIO errors.
* Fix miscalculated number of end frame pulses for increased performance.
* Replace start frame and end frame loops with slices for increased performance.

## 0.3.0 (March 16, 2018)

* Add support for Raspberry Pi 3 B+.

## 0.2.0 (October 6, 2017)

* Update internal struct/enum names for `rppal` 0.2.0 upgrade.
* (Breaking change) Rename `GPIOError` to `GpioError`, and `Error::GPIO` to `Error::Gpio`.

## 0.1.2 (March 1, 2017)

* Move `GPIO` and `System` modules to external crate.
* Remove temporary `blinkt` variable binding in `new()`.

## 0.1.1 (September 6, 2016)

* Add start frame to `show()`.

## 0.1.0 (September 2, 2016)

* Initial release.


src/pixel.rs
const DEFAULT_BRIGHTNESS: u8 = 7;

const IDX_BRIGHTNESS: usize = 0;
const IDX_BLUE: usize = 1;
const IDX_GREEN: usize = 2;
const IDX_RED: usize = 3;

/// A pixel on an LED strip or board.
#[derive(Debug, Copy, Clone)]
pub struct Pixel {
    value: [u8; 4], // Brightness, blue, green, red
}

impl Pixel {
    /// Returns a tuple containing the values for red, green and blue.
    #[inline]
    pub fn rgb(&self) -> (u8, u8, u8) {
        (
            self.value[IDX_RED],
            self.value[IDX_GREEN],
            self.value[IDX_BLUE],
        )
    }

    /// Sets the values for red, green and blue.
    ///
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    #[inline]
    pub fn set_rgb(&mut self, red: u8, green: u8, blue: u8) {
        self.value[IDX_RED] = red;
        self.value[IDX_GREEN] = green;
        self.value[IDX_BLUE] = blue;
    }

    /// Returns a tuple containing the values for red, green, blue and brightness.
    #[inline]
    pub fn rgbb(&self) -> (u8, u8, u8, f32) {
        (
            self.value[IDX_RED],
            self.value[IDX_GREEN],
            self.value[IDX_BLUE],
            f32::from(0b0001_1111 & self.value[IDX_BRIGHTNESS]) / 31.0,
        )
    }

    /// Sets the values for red, green, blue and brightness.
    ///
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    #[inline]
    pub fn set_rgbb(&mut self, red: u8, green: u8, blue: u8, brightness: f32) {
        self.set_rgb(red, green, blue);
        self.set_brightness(brightness);
    }

    /// Returns the red value.
    #[inline]
    pub fn red(&self) -> u8 {
        self.value[IDX_RED]
    }

    /// Sets the red value.
    ///
    /// `red` is specified as an 8-bit value between `0` (0%) and `255` (100%).
    #[inline]
    pub fn set_red(&mut self, red: u8) {
        self.value[IDX_RED] = red;
    }

    /// Returns the green value.
    #[inline]
    pub fn green(&self) -> u8 {
        self.value[IDX_GREEN]
    }

    /// Sets the green value.
    ///
    /// `green` is specified as an 8-bit value between `0` (0%) and `255` (100%).
    #[inline]
    pub fn set_green(&mut self, green: u8) {
        self.value[IDX_GREEN] = green;
    }

    /// Returns the blue value.
    #[inline]
    pub fn blue(&self) -> u8 {
        self.value[IDX_BLUE]
    }

    /// Sets the blue value.
    ///
    /// `blue` is specified as an 8-bit value between `0` (0%) and `255` (100%).
    #[inline]
    pub fn set_blue(&mut self, blue: u8) {
        self.value[IDX_BLUE] = blue;
    }

    /// Returns the brightness value.
    #[inline]
    pub fn brightness(&self) -> f32 {
        f32::from(0b0001_1111 & self.value[IDX_BRIGHTNESS]) / 31.0
    }

    /// Sets the brightness value.
    ///
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    #[inline]
    pub fn set_brightness(&mut self, brightness: f32) {
        self.value[IDX_BRIGHTNESS] = 0b1110_0000 | ((31.0 * brightness.max(0.0).min(1.0)) as u8);
    }

    /// Sets the red, green and blue values to `0`.
    #[inline]
    pub fn clear(&mut self) {
        self.set_rgb(0, 0, 0);
    }

    #[inline]
    pub(crate) fn bytes(&self) -> &[u8] {
        &self.value
    }
}

impl Default for Pixel {
    fn default() -> Self {
        Self {
            value: [0b1110_0000 | DEFAULT_BRIGHTNESS, 0, 0, 0],
        }
    }
}


src/system.rs
// Copyright (c) 2016 Rene van der Meer
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

//! Identifies the Raspberry Pi model based on `/proc/cpuinfo`.

use std::fs::File;
use std::io::{BufReader, BufRead};
use std::result;

const BCM2708_PERIPHERAL_BASE: u32 = 0x20000000;
const BCM2709_PERIPHERAL_BASE: u32 = 0x3f000000;
const GPIO_OFFSET: u32 = 0x200000;

quick_error! {
    #[derive(Debug)]
/// Errors that can occur when trying to identify the Raspberry Pi model.
    pub enum Error {
        UnknownSoC { description("unknown BCM SoC") }
        UnknownModel { description("unknown Raspberry Pi model") }
        CantAccessProcCPUInfo { description("can't access /proc/cpuinfo") }
    }
}

/// Result type returned from methods that can have `system::Error`s.
pub type Result<T> = result::Result<T, Error>;

#[derive(Debug, PartialEq)]
pub enum Model {
    RaspberryPiA,
    RaspberryPiAPlus,
    RaspberryPiB,
    RaspberryPiBPlus,
    RaspberryPi2B,
    RaspberryPi3B,
    RaspberryPiCompute,
    RaspberryPiZero,
    Unknown,
}

#[derive(Debug, PartialEq)]
pub enum SoC {
    BCM2835,
    BCM2836,
    BCM2837,
}

pub struct System {
    pub model: Model,
    pub soc: SoC,
    pub peripheral_base: u32,
    pub gpio_offset: u32,
}

impl System {
    pub fn new() -> Result<System> {
        // Parse hardware/revision from /proc/cpuinfo to figure out model/SoC
        let proc_cpuinfo = BufReader::new(match File::open("/proc/cpuinfo") {
            Err(_) => return Err(Error::CantAccessProcCPUInfo),
            Ok(file) => file,
        });

        let mut hardware: String = String::new();
        let mut revision: String = String::new();
        for line_result in proc_cpuinfo.lines() {
            if let Some(line) = line_result.ok() {
                if line.starts_with("Hardware\t: ") {
                    hardware = String::from(&line[11..]);
                } else if line.starts_with("Revision\t: ") {
                    revision = String::from(&line[11..]);
                }
            }
        }

        // Make sure we're actually running on a compatible BCM SoC
        match &hardware[..] {
            "BCM2708" | "BCM2709" | "BCM2835" | "BCM2836" | "BCM2837" => {}
            _ => return Err(Error::UnknownSoC),
        }

        let mut model = Model::Unknown;

        // Newer revisions consist of at least 6 characters
        if revision.len() >= 6 {
            model = match &revision[revision.len() - 3..revision.len() - 1] {
                "00" => Model::RaspberryPiA,
                "01" => Model::RaspberryPiB,
                "02" => Model::RaspberryPiAPlus,
                "03" => Model::RaspberryPiBPlus,
                "04" => Model::RaspberryPi2B,
                "06" => Model::RaspberryPiCompute,
                "08" => Model::RaspberryPi3B,
                "09" => Model::RaspberryPiZero,
                _ => Model::Unknown,
            };
        } else if revision.len() == 4 {
            // Older revisions are 4 characters long
            model = match &revision[..] {
                "0007" | "0008" | "0009" => Model::RaspberryPiA,
                "0002" | "0003" | "0004" | "0005" | "0006" | "000d" | "000e" | "000f" => {
                    Model::RaspberryPiB
                }
                "0012" => Model::RaspberryPiAPlus,
                "0010" | "0013" => Model::RaspberryPiBPlus,
                "0011" => Model::RaspberryPiCompute,
                _ => Model::Unknown,
            };
        }

        match model {
            Model::RaspberryPiA |
            Model::RaspberryPiAPlus |
            Model::RaspberryPiB |
            Model::RaspberryPiBPlus |
            Model::RaspberryPiCompute |
            Model::RaspberryPiZero => {
                Ok(System {
                    model: model,
                    soc: SoC::BCM2835,
                    peripheral_base: BCM2708_PERIPHERAL_BASE,
                    gpio_offset: GPIO_OFFSET,
                })
            }
            Model::RaspberryPi2B => {
                Ok(System {
                    model: model,
                    soc: SoC::BCM2836,
                    peripheral_base: BCM2709_PERIPHERAL_BASE,
                    gpio_offset: GPIO_OFFSET,
                })
            }
            Model::RaspberryPi3B => {
                Ok(System {
                    model: model,
                    soc: SoC::BCM2837,
                    peripheral_base: BCM2709_PERIPHERAL_BASE,
                    gpio_offset: GPIO_OFFSET,
                })
            }
            _ => Err(Error::UnknownModel),
        }
    }
}


examples/random.rs
// random.rs - Sets each pixel on a Blinkt! board to a random RGB value in a loop.
//
// Interrupting the process by pressing Ctrl-C causes the application to exit
// immediately without clearing the pixels. Check out the solid_signals.rs
// example to learn how to properly handle incoming signals to prevent an
// abnormal termination.

use std::error::Error;
use std::thread;
use std::time::Duration;

use rand::{self, Rng};

use blinkt::Blinkt;

fn main() -> Result<(), Box<dyn Error>> {
    let mut blinkt = Blinkt::new()?;
    let mut rng = rand::thread_rng();

    blinkt.set_all_pixels_brightness(0.1);

    loop {
        // Iterate over all pixels, setting red, green and blue to random values.
        for pixel in &mut blinkt {
            pixel.set_rgb(rng.gen(), rng.gen(), rng.gen());
        }

        // Send the new color values to the Blinkt! board.
        blinkt.show()?;

        thread::sleep(Duration::from_millis(100));
    }
}


src/lib.rs
//! A Rust library that provides an interface for the Pimoroni Blinkt!, and any
//! similar APA102 or SK9822 LED strips or boards, on a Raspberry Pi. The library
//! supports bitbanging mode on any GPIO pins, and hardware SPI mode on GPIO 10
//! (physical pin 19) for data, and GPIO 11 (physical pin 23) for clock.
//!
//! For bitbanging mode, Blinkt gains access to the BCM283x GPIO peripheral either
//! through `/dev/gpiomem` or `/dev/mem`. Hardware SPI mode is controlled
//! through `/dev/spidev0.0`.
//!
//! Both the original APA102 and the SK9822 clone are supported. The APA102 RGB
//! LED/driver ICs are referred to as pixels throughout the code and documentation.
//!
//! Each pixel has a red, green and blue LED with possible values between 0-255.
//! Additionally, the overall brightness of each pixel can be set to 0.0-1.0, which
//! is converted to a 5-bit value.
//!
//! Blinkt stores all color and brightness changes in a local buffer. Use
//! `show()` to send the buffered values to the pixels.
//!
//! By default, all pixels are cleared when Blinkt goes out of
//! scope. Use `set_clear_on_drop(false)` to disable this behavior. Note that
//! `drop` methods aren't called when a process is abnormally terminated (for
//! instance when a `SIGINT` isn't caught).
//!
//! # Examples
//!
//! ### Blinkt! board
//!
//! A complete example that cycles all pixels on a Blinkt! board through red, green
//! and blue.
//!
//! ```rust,no_run
//! use std::error::Error;
//! use std::time::Duration;
//! use std::{thread, mem};
//!
//! use blinkt::Blinkt;
//!
//! fn main() -> Result<(), Box<dyn Error>> {
//!     let mut blinkt = Blinkt::new()?;
//!     let (red, green, blue) = (&mut 255, &mut 0, &mut 0);
//!
//!     loop {
//!         blinkt.set_all_pixels(*red, *green, *blue);
//!         blinkt.show()?;
//!
//!         thread::sleep(Duration::from_millis(250));
//!
//!         mem::swap(red, green);
//!         mem::swap(red, blue);
//!     }
//! }
//! ```
//!
//! ### APA102 or SK9822 LED strip
//!
//! The recommended way to control an LED strip is to use the hardware SPI
//! interface through `Blinkt::with_spi()`, with the data line connected to GPIO 10
//! (physical pin 19), and clock on GPIO 11 (physical pin 23).
//!
//! ```rust,no_run
//! # use std::error::Error;
//! #
//! # use blinkt::{Blinkt, BlinktSpi};
//! #
//! # fn main() {
//! let mut blinkt = Blinkt::with_spi(BlinktSpi::default(), 144);
//! # }
//! ```
//!
//! Alternatively, you can use the bitbanging mode through `Blinkt::with_settings()`
//! to connect the LED strip to any available GPIO pins. However, this is less reliable
//! than using the hardware SPI interface, and may cause issues on longer strips.
//!
//! ```rust,no_run
//! # use std::error::Error;
//! #
//! # use blinkt::Blinkt;
//! #
//! # fn main() -> Result<(), Box<dyn Error>> {
//! let mut blinkt = Blinkt::with_settings(23, 24, 8)?;
//! # Ok(())
//! # }
//! ```
//!
//! ### Iterators
//!
//! Besides the various `set_` methods available on `Blinkt`, pixels can also be accessed
//! and modified through an iterator. `Blinkt::iter_mut()` returns a mutable iterator over
//! all `Pixel`s stored in `Blinkt`.
//!
//! ```rust,no_run
//! # use std::error::Error;
//! #
//! # use blinkt::Blinkt;
//! #
//! # fn main() -> Result<(), Box<dyn Error>> {
//! #     let mut blinkt = Blinkt::new()?;
//! #
//! blinkt.iter_mut().for_each(|pixel| {
//!     pixel.set_rgb(255, 0, 255);
//! });
//! #    Ok(())
//! # }
//! ```
//!
//! For more idiomatic `for` loops, you can access the same iterator
//! through a mutable reference of a `Blinkt` instance.
//!
//! ```rust,no_run
//! # use std::error::Error;
//! #
//! # use blinkt::Blinkt;
//! #
//! # fn main() -> Result<(), Box<dyn Error>> {
//! #     let mut blinkt = Blinkt::new()?;
//! #
//! for pixel in &mut blinkt {
//!     pixel.set_rgb(255, 255, 0);
//! }
//! #    Ok(())
//! # }
//! ```
//!
// Used by rustdoc to link other crates to blinkt's docs
#![doc(html_root_url = "https://docs.rs/blinkt/0.7.1")]
#![allow(clippy::trivially_copy_pass_by_ref)]

use std::error;
use std::fmt;
use std::io;
use std::result;
use std::slice;
use std::time::Duration;

use rppal::gpio::{Gpio, OutputPin};

pub use rppal::gpio::Error as GpioError;
pub use rppal::spi::Error as SpiError;

mod pixel;

pub use pixel::Pixel;

// Default values for the Pimoroni Blinkt! board using BCM GPIO pin numbers
const DAT: u8 = 23;
const CLK: u8 = 24;
const NUM_PIXELS: usize = 8;

#[derive(Debug)]
/// Errors that can occur while using Blinkt.
pub enum Error {
    /// Accessing the GPIO peripheral returned an error.
    ///
    /// Some of these errors can be fixed by changing file permissions, or upgrading
    /// to a more recent version of Raspbian.
    Gpio(GpioError),
    /// Accessing the SPI peripheral returned an error.
    Spi(SpiError),
    /// An I/O operation returned an error.
    Io(io::Error),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Error::Gpio(ref err) => write!(f, "GPIO error: {}", err),
            Error::Spi(ref err) => write!(f, "SPI error: {}", err),
            Error::Io(ref err) => write!(f, "I/O error: {}", err),
        }
    }
}

impl error::Error for Error {}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self {
        Self::Io(err)
    }
}

impl From<GpioError> for Error {
    fn from(err: GpioError) -> Self {
        Self::Gpio(err)
    }
}

impl From<SpiError> for Error {
    fn from(err: SpiError) -> Self {
        Self::Spi(err)
    }
}

/// Result type returned from methods that can have `blinkt::Error`s.
pub type Result<T> = result::Result<T, Error>;

trait SerialOutput {
    fn write(&mut self, data: &[u8]) -> Result<()>;
}

struct BlinktGpio {
    pin_data: OutputPin,
    pin_clock: OutputPin,
}

impl BlinktGpio {
    pub fn with_settings(pin_data: u8, pin_clock: u8) -> Result<Self> {
        let gpio = Gpio::new()?;

        let mut pin_data = gpio.get(pin_data)?.into_output();
        let mut pin_clock = gpio.get(pin_clock)?.into_output();

        pin_data.set_low();
        pin_clock.set_low();

        Ok(Self {
            pin_data,
            pin_clock,
        })
    }
}

impl SerialOutput for BlinktGpio {
    fn write(&mut self, data: &[u8]) -> Result<()> {
        for byte in data {
            for n in 0..8 {
                if (byte & (1 << (7 - n))) > 0 {
                    self.pin_data.set_high();
                } else {
                    self.pin_data.set_low();
                }

                self.pin_clock.set_high();
                std::thread::sleep(Duration::from_nanos(10000));
                self.pin_clock.set_low();
            }
        }

        Ok(())
    }
}

pub mod spi {
    pub(crate) use rppal::spi::Spi;
    pub use rppal::spi::{Bus, Mode, SlaveSelect};
}

pub struct BlinktSpi(spi::Spi);

impl BlinktSpi {
    pub fn with_settings(
        bus: spi::Bus,
        slave: spi::SlaveSelect,
        clock_speed_hz: u32,
        mode: spi::Mode,
    ) -> Result<Self> {
        Ok(Self(spi::Spi::new(bus, slave, clock_speed_hz, mode)?))
    }
}

impl Default for BlinktSpi {
    fn default() -> Self {
        Self(
            spi::Spi::new(
                spi::Bus::Spi0,
                spi::SlaveSelect::Ss0,
                1_000_000,
                spi::Mode::Mode0,
            )
            .expect("Can't create spi bus"),
        )
    }
}

impl SerialOutput for BlinktSpi {
    fn write(&mut self, data: &[u8]) -> Result<()> {
        self.0.write(data)?;

        Ok(())
    }
}

/// Interface for the Pimoroni Blinkt!, and any similar APA102 or SK9822 LED
/// strips or boards.
///
/// By default, Blinkt is set up to communicate with an 8-pixel board through
/// data pin GPIO 23 (physical pin 16) and clock pin GPIO 24 (physical pin 18).
/// These settings can be changed to support alternate configurations.
pub struct Blinkt {
    serial_output: Box<dyn SerialOutput + Send>,
    pixels: Vec<Pixel>,
    clear_on_drop: bool,
    end_frame: Vec<u8>,
}

impl Blinkt {
    /// Constructs a new `Blinkt` using the default settings for a Pimoroni
    /// Blinkt! board.
    ///
    /// This sets the data pin to GPIO 23 (physical pin 16), the clock pin to
    /// GPIO 24 (physical pin 18), and number of pixels to 8.
    pub fn new() -> Result<Self> {
        Self::with_settings(DAT, CLK, NUM_PIXELS)
    }

    /// Constructs a new `Blinkt` using bitbanging mode, with custom settings for
    /// the data pin, clock pin, and number of pixels. Pins should be specified
    /// by their BCM GPIO pin numbers.
    pub fn with_settings(pin_data: u8, pin_clock: u8, num_pixels: usize) -> Result<Self> {
        Ok(Self {
            serial_output: Box::new(BlinktGpio::with_settings(pin_data, pin_clock)?),
            pixels: vec![Pixel::default(); num_pixels],
            clear_on_drop: true,
            end_frame: vec![0u8; 4 + (((num_pixels as f32 / 16.0f32) + 0.94f32) as usize)],
        })
    }

    /// Constructs a new `Blinkt` using hardware SPI, with custom settings for the
    /// clock speed and number of pixels.
    ///
    /// This sets the data pin to GPIO 10 (physical pin 19) and the clock pin
    /// to GPIO 11 (physical pin 23).
    ///
    /// The Raspberry Pi allows SPI clock speeds up to 125 MHz (125_000_000),
    /// but the maximum speed supported by LED strips depends a lot on the
    /// number of pixels and wire quality, and requires some experimentation.
    /// 32 MHz (32_000_000) seems to be the maximum clock speed for a typical
    /// short LED strip. Visit the [Raspberry Pi SPI Documentation](https://www.raspberrypi.org/documentation/hardware/raspberrypi/spi/)
    /// page for a complete list of supported clock speeds.
    pub fn with_spi(spi: BlinktSpi, num_pixels: usize) -> Self {
        Self {
            serial_output: Box::new(spi),
            pixels: vec![Pixel::default(); num_pixels],
            clear_on_drop: true,
            end_frame: vec![0u8; 4 + (((num_pixels as f32 / 16.0f32) + 0.94f32) as usize)],
        }
    }

    /// Returns a mutable iterator over all `Pixel`s stored in `Blinkt`.
    pub fn iter_mut(&mut self) -> slice::IterMut<'_, Pixel> {
        self.pixels.iter_mut()
    }

    /// Sets the red, green and blue values for a single pixel in the local
    /// buffer.
    ///
    /// Pixels are numbered starting at `0`.
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    pub fn set_pixel(&mut self, pixel: usize, red: u8, green: u8, blue: u8) {
        if let Some(pixel) = self.pixels.get_mut(pixel) {
            pixel.set_rgb(red, green, blue);
        }
    }

    /// Sets the red, green, blue and brightness values for a single pixel in
    /// the local buffer.
    ///
    /// Pixels are numbered starting at `0`.
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    pub fn set_pixel_rgbb(&mut self, pixel: usize, red: u8, green: u8, blue: u8, brightness: f32) {
        if let Some(pixel) = self.pixels.get_mut(pixel) {
            pixel.set_rgbb(red, green, blue, brightness);
        }
    }

    /// Sets the brightness value for a single pixel in the local buffer.
    ///
    /// Pixels are numbered starting at `0`.
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    pub fn set_pixel_brightness(&mut self, pixel: usize, brightness: f32) {
        if let Some(pixel) = self.pixels.get_mut(pixel) {
            pixel.set_brightness(brightness);
        }
    }

    /// Sets the red, green and blue values for all pixels in the local buffer.
    ///
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    pub fn set_all_pixels(&mut self, red: u8, green: u8, blue: u8) {
        for pixel in &mut self.pixels {
            pixel.set_rgb(red, green, blue);
        }
    }

    /// Sets the red, green, blue and brightness values for all pixels in the
    /// local buffer.
    ///
    /// `red`, `green` and `blue` are specified as 8-bit values between `0` (0%) and `255` (100%).
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    pub fn set_all_pixels_rgbb(&mut self, red: u8, green: u8, blue: u8, brightness: f32) {
        for pixel in &mut self.pixels {
            pixel.set_rgbb(red, green, blue, brightness);
        }
    }

    /// Sets the brightness value for all pixels.
    ///
    /// `brightness` is specified as a floating point value between `0.0` (0%) and `1.0` (100%), and is converted to a 5-bit value.
    pub fn set_all_pixels_brightness(&mut self, brightness: f32) {
        for pixel in &mut self.pixels {
            pixel.set_brightness(brightness);
        }
    }

    /// Sets the red, green and blue values for all pixels to `0`.
    pub fn clear(&mut self) {
        self.set_all_pixels(0, 0, 0);
    }

    /// Sends the contents of the local buffer to the pixels, updating their
    /// LED colors and brightness.
    pub fn show(&mut self) -> Result<()> {
        // Start frame (32*0).
        self.serial_output.write(&[0u8; 4])?;

        // LED frames (3*1, 5*brightness, 8*blue, 8*green, 8*red).
        for pixel in &self.pixels {
            self.serial_output.write(pixel.bytes())?;
        }

        // End frame (8*0 for every 16 pixels, 32*0 SK9822 reset frame).
        // The SK9822 won't update any pixels until it receives the next
        // start frame (32*0). The APA102 doesn't care if we send zeroes
        // instead of ones as the end frame. This workaround is
        // compatible with both the APA102 and SK9822.
        self.serial_output.write(&self.end_frame)?;

        Ok(())
    }

    /// Returns the value of `clear_on_drop`.
    pub fn clear_on_drop(&self) -> bool {
        self.clear_on_drop
    }

    /// When enabled, clears all pixels when `Blinkt` goes out of scope.
    ///
    /// By default, this is set to `true`.
    ///
    /// ## Note
    ///
    /// Drop methods aren't called when a process is abnormally terminated, for
    /// instance when a user presses <kbd>Ctrl</kbd> + <kbd>C</kbd>, and the `SIGINT` signal
    /// isn't caught. You can catch those using crates such as [`simple_signal`].
    ///
    /// [`simple_signal`]: https://crates.io/crates/simple-signal
    pub fn set_clear_on_drop(&mut self, clear_on_drop: bool) {
        self.clear_on_drop = clear_on_drop;
    }
}

impl Drop for Blinkt {
    /// Clears all pixels if [`clear_on_drop`] is set to `true` (default).
    ///
    /// [`clear_on_drop`]: #method.clear_on_drop
    fn drop(&mut self) {
        if self.clear_on_drop {
            self.clear();
            let _ = self.show();
        }
    }
}

impl<'a> IntoIterator for &'a mut Blinkt {
    type Item = &'a mut Pixel;
    type IntoIter = slice::IterMut<'a, Pixel>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter_mut()
    }
}
